stumpy/gpu_stump.py

```diff
@@ -621,22 +829,44 @@ def gpu_stump(
 
     for idx in range(len(device_ids)):
         profile_fname = profile[idx]
+        profile_L_fname = profile_L[idx]
+        profile_R_fname = profile_R[idx]
         indices_fname = indices[idx]
+        indices_L_fname = indices_L[idx]
+        indices_R_fname = indices_R[idx]
+
         profile[idx] = np.load(profile_fname, allow_pickle=False)
+        profile_L[idx] = np.load(profile_L_fname, allow_pickle=False)
+        profile_R[idx] = np.load(profile_R_fname, allow_pickle=False)
         indices[idx] = np.load(indices_fname, allow_pickle=False)
+        indices_L[idx] = np.load(indices_L_fname, allow_pickle=False)
+        indices_R[idx] = np.load(indices_R_fname, allow_pickle=False)
+
         os.remove(profile_fname)
+        os.remove(profile_L_fname)
+        os.remove(profile_R_fname)
         os.remove(indices_fname)
-
-    for i in range(1, len(device_ids)):
-        # Update all matrix profiles and matrix profile indices
-        # (global, left, right) and store in profile[0] and indices[0]
-        for col in range(profile[0].shape[1]):  # pragma: no cover
-            cond = profile[0][:, col] < profile[i][:, col]
-            profile[0][:, col] = np.where(cond, profile[0][:, col], profile[i][:, col])
-            indices[0][:, col] = np.where(cond, indices[0][:, col], indices[i][:, col])
-
-    out[:, 0] = profile[0][:, 0]
-    out[:, 1:4] = indices[0][:, :]
+        os.remove(indices_L_fname)
+        os.remove(indices_R_fname)
+
+    for i in range(1, len(device_ids)):  # pragma: no cover
+        # Update (top-k) matrix profile and matrix profile indices
+        core._merge_topk_PI(profile[0], profile[i], indices[0], indices[i])
+
+        # Update (top-1) left matrix profile and matrix profile indices
+        cond = profile_L[0] < profile_L[i]
```

Comment:
So, `profile_L` is defined as a nested lists: `profile_L = [None] * len(device_ids)`. It seems it should be okay to change it to numpy array: `profile_L = np.empty((l, len(device_ids), dtype=np.float64)`

And, for `profile`, it would be: `profile = np.empty((l, k, len(device_ids), dtype=np.float64)`. Is that okay to change them to numpy array? or, is there a specific reason?

Reply:
> ask = profile_L[0] < profile_L[i]
profile_L[0][mask] = profile_L[i][mask]
indices_L[0][mask] = indices_L[i][mask]

Yes, exactly! I did, but then I remember you do not like chain slicing. So, I tried to see if I can make it numpy array.

> Note that `_gpu_stump` doesn't actually return arrays. It returns a file name when then eventually gets read into memory and converted into a numpy array

I didn't pay attention to that! Thanks for letting me know!